File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque
def solution(n, vertex):
    
    node = [[] for _ in range(n+1)]
    visited = [0] * (n+1)
    
    for a, b in vertex:
        node[a].append(b)
        node[b].append(a)
        
    visited[1]=1
    q = deque([1])
    
    while q:
        curr = q.popleft()
        for next in node[curr]:
            if not visited[next]:
                visited[next] = visited[curr]+1
                q.append(next)
    
    max_v = max(visited)
    cnt = visited.count(max_v)
        
    return cnt if cnt >0 else 1

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    
    let graph = Array.from({length:n+1},()=>[])
    for(let [s,e] of edge){
        graph[s].push(e)
        graph[e].push(s)
    }
    
    let maxDist = 0
    let visited = Array(n+1).fill(0)
    visited[1] = 1
    
    let queue = []
    let begin = 0
    let end = 1
    queue.push([1]) // node
    
    while(begin < end){
        const curr = queue[begin++]
        
        maxDist = Math.max(maxDist, visited[curr])
        
        for(let next of graph[curr]){
            if(visited[next]==0){
                visited[next] = visited[curr] + 1;
                queue.push([next])
                end++;
            }
        }
    }
    
    let maxCnt = 0
    visited.sort((a,b)=>b-a)
    for(let num of visited){
        if(visited[0]===num) {
            maxCnt++
        }
        else{ 
            break
        }
    }

    return maxCnt;
}
